feat: Add a parallel JSON log by default

favilo opened the pull request:
For ease of programmatic parsing, and uploading to Kibana with filebeat

<!--
Thank you for your interest in and contributing to Rally! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

* Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
* Have you followed the [contributor guidelines](https://github.com/elastic/rally/blob/master/CONTRIBUTING.md)?
* Have you run `make check-all` successfully?
* Did you choose a [descriptive title and description](https://chris.beams.io/posts/git-commit/) for your PR?
* (Only for maintainers) Did you apply appropriate labels and a milestone?

gbanasiak commented:
Need to merge https://github.com/elastic/rally/pull/1890 in for linter to pass.

gbanasiak reviewed (commented):
I did a quick functional test and I'm not getting timestamp field in the logs:

```
% esrally list races

% cat ~/.rally/logs/rally.log 
2024-11-18 13:12:31,530 -not-actor-/PID:16574 esrally.rally INFO OS [uname_result(system='Darwin', node='pc-913.home', release='23.6.0', version='Darwin Kernel Version 23.6.0: Thu Sep 12 23:35:10 PDT 2024; root:xnu-10063.141.1.701.1~1/RELEASE_ARM64_T6030', machine='arm64')]
2024-11-18 13:12:31,530 -not-actor-/PID:16574 esrally.rally INFO Python [namespace(name='cpython', cache_tag='cpython-312', version=sys.version_info(major=3, minor=12, micro=2, releaselevel='final', serial=0), hexversion=51118832, _multiarch='darwin')]
2024-11-18 13:12:31,557 -not-actor-/PID:16574 esrally.rally INFO Rally version [2.11.1.dev0 (git revision: 066bb7587468b4bef74b3b5978042d1dfd074d97)]
2024-11-18 13:12:31,557 -not-actor-/PID:16574 esrally.utils.net INFO Connecting directly to the Internet (no proxy support) for [all_proxy].
2024-11-18 13:12:31,557 -not-actor-/PID:16574 esrally.utils.net INFO Connecting directly to the Internet (no proxy support) for [all_proxy].
2024-11-18 13:12:31,557 -not-actor-/PID:16574 esrally.rally INFO Cleaning track dependency directory [/Users/grzegorz/.rally/libs]...
2024-11-18 13:12:31,557 -not-actor-/PID:16574 esrally.metrics INFO Creating file race store

% cat ~/.rally/logs/rally.json
{"message": "OS [uname_result(system='Darwin', node='pc-913.home', release='23.6.0', version='Darwin Kernel Version 23.6.0: Thu Sep 12 23:35:10 PDT 2024; root:xnu-10063.141.1.701.1~1/RELEASE_ARM64_T6030', machine='arm64')]", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Python [namespace(name='cpython', cache_tag='cpython-312', version=sys.version_info(major=3, minor=12, micro=2, releaselevel='final', serial=0), hexversion=51118832, _multiarch='darwin')]", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Rally version [2.11.1.dev0 (git revision: 066bb7587468b4bef74b3b5978042d1dfd074d97)]", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Connecting directly to the Internet (no proxy support) for [all_proxy].", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Connecting directly to the Internet (no proxy support) for [all_proxy].", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Cleaning track dependency directory [/Users/grzegorz/.rally/libs]...", "taskName": null, "actorAddress": "-not-actor-"}
{"message": "Creating file race store", "taskName": null, "actorAddress": "-not-actor-"}

% python --version
Python 3.12.2

% pip list | grep python-json
python-json-logger            2.0.7
```

favilo commented:
> I did a quick functional test and I'm not getting timestamp field in the logs:

Care to give it a try again @gbanasiak? I've added the `timestamp` field to the `LogRecord`. It will correspond to the `%(asctime)s,%(msecs)d` that we've been using for the other formatters. And I've added the ability to specify `timezone`

gbanasiak commented:
That's better, now I'm getting this after formatting:

```
{
    "timestamp": "2024-11-19 11:09:55,605074",
    "process": 12826,
    "name": "esrally.driver.driver",
    "levelname": "INFO",
    "message": "Worker[0] finished executing tasks ['sort_country_code_no_can_match_shortcut'] in 0.020128 seconds",
    "taskName": "Task-145",
    "actorAddress": "ActorAddr-(T|:51749)",
    "timezone": "UTC"
}
```

There's a new `taskName` log attribute injected by Async I/O which is not included in plain logs. Nice.

To make this easier to ingest by Filebeat / Elastic Agent we could change `timestamp` to `@timestamp` and adjust format to be ISO compatible, otherwise we would get error in ES:
```
failed to parse date field [2024-11-19 11:09:55,605074] with format [strict_date_optional_time||epoch_millis]
```
This could be addressed by proper Filebeat configuration too.

I was thinking about taking this one step further and modifying log attributes to be [ECS](https://www.elastic.co/guide/en/ecs/current/index.html) compliant (also [ECS fields](https://www.elastic.co/guide/en/beats/filebeat/current/exported-fields-ecs.html) in Filebeat docs). Perhaps this could be done with [object translator](https://github.com/madzak/python-json-logger?tab=readme-ov-file#adding-custom-object-serialization)? But I don't find obvious candidates for some of the fields:

```
timestamp -> @timestamp
process -> process.pid
name - ?
levelname - log.level
message - message
taskName - ?
actorAddress - ?
timezone - not needed if @timestamp properly formatted
```

gbanasiak commented:
> I was thinking about taking this one step further and modifying log attributes to be [ECS](https://www.elastic.co/guide/en/ecs/current/index.html) compliant [..]

We have to change `process` field name to something different due to collision with Filebeat mappings. The fields that have no good candidates could be placed under `rally` top-level key.

favilo commented:
I've used the `ecs-logging` python library to actually implement on-spec JSON logging. This is much better than my adhoc JSON formatter from before. 

I've also subclassed the formatter in order to move the thespian log fields into their own `rally.actor` section.

gbanasiak commented:
Example result:

```
{
  "@timestamp": "2024-11-20T09:39:43.312Z",
  "log.level": "warning",
  "message": "Invoking a scroll search with the 'search' operation is deprecated and will be removed in a future release. Use 'scroll-search' instead.",
  "ecs.version": "1.6.0",
  "log": {
    "logger": "esrally.driver.runner",
    "origin": {
      "file": {
        "line": 1214,
        "name": "runner.py"
      },
      "function": "__call__"
    },
    "original": "Invoking a scroll search with the 'search' operation is deprecated and will be removed in a future release. Use 'scroll-search' instead."
  },
  "process": {
    "name": "Actor_Worker__ActorAddr-LocalAddr.1",
    "pid": 30676,
    "thread": {
      "id": 130507456198208,
      "name": "ThreadPoolExecutor-0_0"
    }
  },
  "rally": {
    "actor": {
      "address": "ActorAddr-(T|:45685)"
    }
  }
}
```

gbanasiak reviewed (commented):
Great idea to use `ecs-logging`. I did another round of e2e tests and found no collisions when ingesting with Filebeat.

Can we drop `log.original` entirely? It's useful when documents are processed with ES ingest pipeline, but I don't think we need this (at least not yet) as `message` field has exactly the same content.

I left some comments and questions.

gbanasiak commented on esrally/log.py:
```diff
@@ -40,6 +43,36 @@ def configure_utc_formatter(*args, **kwargs):
+
+    def rename_actor_fields(self, log_record: typing.Dict[str, typing.Any]) -> None:
+        actor = {}
+        if log_record.get("actorAddress"):
+            actor["address"] = log_record.pop("actorAddress")
+        if log_record.get("taskName"):
+            actor["task"] = log_record.pop("taskName")
+        if actor:
+            collections.deep_update(log_record, {"rally": {"actor": actor}})
```
Or we can just drop `taskName` for now.

gbanasiak replied:
> Huh, turns out that's fairly new.

Ah, yes, most recently I've tested with 3.11.7 from esbench env. After switching to 3.12.2 and:

```
diff --git a/esrally/resources/logging.json b/esrally/resources/logging.json
index c1e4b6da..7651ebc5 100644
--- a/esrally/resources/logging.json
+++ b/esrally/resources/logging.json
@@ -2,12 +2,12 @@
   "version": 1,
   "formatters": {
     "normal": {
-      "format": "%(asctime)s,%(msecs)d %(actorAddress)s/PID:%(process)d %(name)s %(levelname)s %(message)s",
+      "format": "%(asctime)s,%(msecs)d %(actorAddress)s/PID:%(process)d %(name)s %(taskName)s %(levelname)s %(message)s",
       "datefmt": "%Y-%m-%d %H:%M:%S",
       "()": "esrally.log.configure_utc_formatter"
     },
     "json": {
-      "format": "%(message)s",
+      "format": "%(message)s %(taskName)s",
       "()": "esrally.log.configure_ecs_formatter"
     },
     "profile": {
```

I can see `taskName` populated in `rally.log` but not in `rally.json`. I think it gets swallowed in [here](https://github.com/elastic/ecs-logging-python/blob/a4ef6e1ffff4f41b0f66d52a8dc93e77f6f70ac3/ecs_logging/_stdlib.py#L167). We could override `format_to_ecs()` but I'm also fine just skipping it for now. We can revisit once we find a use for this information. What we're adding in `process` object is already very useful.

Still interested in hearing the rationale behind `deep_update()` :)

favilo replied:
`deep_update` is so that _just in case_ we have something already there, like someone adds a future mutator, we don't clobber any dict that is already defined. Like would happen with `dict.update()`

favilo commented on docs/configuration.rst:
```diff
@@ -207,7 +209,7 @@ The log file will not be rotated automatically as this is problematic due to Ral
 ~~~~~~~
 
-With the following configuration Rally will log all output to standard error::
+With the following configuration Rally will log all output to standard error, and format the timestamps in the local timezone::
```
It still is for rally.log

favilo commented:
> Can we drop `log.original` entirely?

Absolutely, I'm just going to drop that in `logging.json`, by adding an `exclude_fields` field. That way, of someone wants it back, they can just edit the `logging.json` file

gbanasiak commented on docs/configuration.rst:
```diff
@@ -183,7 +183,9 @@ Logging in Rally is configured in ``~/.rally/logging.json``. For more informatio
 
-By default, Rally will log all output to ``~/.rally/logs/rally.log``. The default timestamp is UTC, but users can opt for the local time by setting ``"timezone": "localtime"`` in the logging configuration file.
+By default, Rally will log all output to ``~/.rally/logs/rally.log`` in plain text format and ``~/.rally/logs/rally.json`` in JSON format.
+The default timestamp for ``rally.log`` is UTC, but users can opt for the local time by setting ``"timezone": "localtime"`` in the logging configuration file. 
```
Ah, yes. I was looking for `timezone` in this change missing the fact it came earlier in https://github.com/elastic/rally/pull/1727.

favilo commented:
Okay!

It's been a while since I remembered this, but I've added the ability to change mutators and such.

The current format:
```{
  "@timestamp":"2024-12-10T20:26:20.561Z",
  "log.level":"info",
  "message":"Hello world",
  "ecs.version":"1.6.0",
  "log":{"logger":"esrally.rally","origin":{"file":{"line":1238,"name":"rally.py"},"function":"test"}},
  "process":{"name":"MainProcess","pid":3061691,"thread":{"id":140008672693120,"name":"MainThread"}},
  "python":{"asyncio":{"task":"Task-1"},
  "thespian":{"address":"-not-actor-"}}}
```

Note: `python.asyncio.task` only shows up if it is running inside an async task. But it does actually show up now. `ecs_logging` was in fact eating that

gbanasiak commented on esrally/log.py:
```diff
@@ -40,6 +43,61 @@ def configure_utc_formatter(*args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.mutators = mutators or []
+
+    def format(self, record: logging.LogRecord) -> str:
+        log_dict = super().format_to_ecs(record)
+        self.apply_mutators(record, log_dict)
+        # ecs_logging._utils is a private module
+        # but we need to use it here to get the on spec JSON serialization
+        return ecs_logging._utils.json_dumps(log_dict)  # pylint: disable=protected-access
```
I think we can avoid protected access like below. This follows [suggestion](https://github.com/elastic/ecs-logging-python/blob/a4ef6e1ffff4f41b0f66d52a8dc93e77f6f70ac3/ecs_logging/_stdlib.py#L168-L179) in `ecs_logging`.

```suggestion
    def format_to_ecs(self, record: logging.LogRecord) -> typing.Dict[str, typing.Any]:
        log_dict = super().format_to_ecs(record)
        self.apply_mutators(record, log_dict)
        return log_dict
```

favilo replied:
Of course I should have done that. Thanks. Not sure where my brain was when I was overriding that method.

gbanasiak reviewed (approved):
Many thanks for adding this. LGTM. See remark regarding protected access.
